feat: Allow passing model and tokenizer to ArgillaTrainer directly

tomaarsen opened the pull request:
Hello!

# Description

Closes #3631.

This is important to give users freedom to very specifically set up their tokenizer. This is required e.g. for SFT with TRL.

**Type of change**

- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing functionality)
- [ ] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

Updated the relevant tests (TRL, Transformers) to also train with the passed model & tokenizer.

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

---

**TODO**:

- [x] CHANGELOG
- [x] Documentation
- [x] Double-check docstrings

---

- Tom Aarsen

tomaarsen committed:
Allow PreTrainedModel and PreTrainedTokenizer for TRL

+ TRL tests

tomaarsen committed:
Allow passing model and tokenizer for transformers

tomaarsen committed:
Speed up transformers test

tomaarsen committed:
Improve TRL tests

tomaarsen committed:
Add note on passing a model and tokenizer to the Trainer

tomaarsen committed:
Add CHANGELOG

davidberenstein1957 commented on docs/_source/guides/llms/practical_guides/fine_tune.md:
````diff
@@ -446,6 +446,12 @@ trainer = ArgillaTrainer(
 ```
 
+```{note}
+You can also initialize the `ArgillaTrainer` with an already initialized `model` and `tokenizer` for additional fine-grained control. This might be useful if you wish to ensure that the tokenizer adds the EOS token.
````
maybe add a hint
```suggestion
You can also initialize the `ArgillaTrainer` with an already initialized `model` and `tokenizer` for additional fine-grained control. This might be useful if you wish to ensure that the tokenizer adds the EOS token. A lack of this token might result in the model generating endlessly.
```

davidberenstein1957 commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -69,6 +72,7 @@ def __init__(
         self._task = task
         self._seed = seed  # split is used for train-test-split and should therefore be fixed
         self._model = model
+        self._tokenizer = tokenizer
```
maybe we should mention if the model and tokenizer passed are not being used in case of other frameworks?

tomaarsen replied:
Good idea

davidberenstein1957 commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -85,16 +86,17 @@ def __init__(
             TrainingTaskForDPO,
```
I think we should update the docstring within the training task to allow for these examples?

tomaarsen replied:
Which docstrings are you referring to? The training task docstrings don't mention the ArgillaTrainer at all

davidberenstein1957 replied:
ohh sorry, maybe the `TrainingTask.for_*`-ones

tomaarsen replied:
Same thing, they only use `dataset.prepare_for_training` in all examples

davidberenstein1957 reviewed (commented):
Looks Good!  some tiny remarks.

tomaarsen committed:
Add hint to doc note

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

tomaarsen committed:
Add warning if tokenizer is passed with the wrong frameworks

Also removed tokenizer from setfit (where it didn't do anything) and
updated some docstrings

tomaarsen committed:
Merge branch 'feat/trainer_model_tokenizer' of https://github.com/argilla-io/argilla into feat/trainer_model_tokenizer

tomaarsen committed:
Merge branch 'develop' into feat/trainer_model_tokenizer
